# Metagenome Assembled Genome Abundance Analysis

This repo uses only an RMarkdown script as input. I followed the following tutorial: https://kbroman.org/blog/2019/02/18/omg_binder/

It plots a heatmap after you mapped reads in your samples against bins or genomes you created. 

 - Launch in RStudio: [![Binder](https://mybinder.org/badge_logo.svg)](https://mybinder.org/v2/gh/megaptera-helvetiae/MAGrepresentation-Binder/master)


## Input 
Example files included:

 - `MetaG-heatmap-Binder.Rmd` - RMarkdown file with R code chunks to plot a heatmap in RStudio
 - `input_matrix_v2.txt` - A sample input file where we mapped reads from several samples against MAG bins that we had created.
 
 
## Output 

Example output includes: 
![alt text](figures/heatmap-relativeAbundance-2-1.png)
![alt text](figures/non-interactive-MAGabund-1.png)

 
# Author ORCID IDs:
- Marian Schmidt: https://orcid.org/0000-0002-2866-4496
- Guillaume Jospin : https://orcid.org/0000-0002-8746-2632
- Laetitia Wilkins: https://orcid.org/0000-0003-3632-2063
- Charles Reid: NA
